tests/Agency/StoreTestAPI.cpp

```diff
@@ -0,0 +1,973 @@
+///
+/// @author Ignacio Rodriguez
+/// @author Max Neunhoeffer
+/// @author Copyright 2021, ArangoDB GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+
+#include <algorithm>
+#include "gtest/gtest.h"
+
+#include "Agency/Store.h"
+#include "Mocks/Servers.h"
+
+#include <velocypack/Builder.h>
+#include <velocypack/Compare.h>
+#include <velocypack/Slice.h>
+#include <velocypack/velocypack-aliases.h>
+
+namespace arangodb {
+namespace tests {
+namespace store_test_api {
+
+class StoreTestAPI : public ::testing::Test {
+ public:
+  StoreTestAPI() : _server(), _store(_server.server(), nullptr) {
+  }
+ protected:  
+  arangodb::tests::mocks::MockCoordinator _server;
+  arangodb::consensus::Store _store;
+
+  auto readAndCheck(std::string const &json) 
+  {
+    try {
+      auto q {VPackParser::fromJson(json)};
+      auto result {std::make_shared<VPackBuilder>()};
+      _store.read(q, result);
+      return result;
+    }
+    catch(std::exception &ex) {
+      throw std::runtime_error(std::string(ex.what()) + " while trying to read " + json);
+    }
+  }
+
+  auto write(std::string const &json)
+  {
+    try {
+      auto q {VPackParser::fromJson(json)};
+      return _store.applyTransactions(q);
+    }
+    catch(std::exception &err) {
+      throw std::runtime_error(std::string(err.what()) + " while parsing " + json );
+    }
+  }
+
+  template<typename TOstream, typename TValueContainer>
+  static TOstream& insert_value_array(TOstream &out, TValueContainer const &src) {
+    out << "[";
+    if (!src.empty()) {
+      std::copy(src.begin(), src.end()-1, std::ostream_iterator<std::string>(out, ", "));
+      out << src.back();
+    }
+    out << "]";
+    return out;
+  }
+
+  auto write(std::vector<std::vector<std::string>> const &operations) {
+    std::stringstream ss;
+    ss << "[";
+    if (!operations.empty()) {
+      insert_value_array(ss, operations.front());
+      for (auto oi {operations.begin() + 1}; oi != operations.end(); ++oi) {
+        insert_value_array(ss, *oi);
+      }
+    }
+    ss << "]";
+    return write(ss.str());
+  }
+
+  auto transactAndCheck(std::string const &json) {
+    auto q {VPackParser::fromJson(json)};
+    auto results { _store.applyTransactions(q)};
+    return results;
+  }
+
+  auto writeAndCheck(std::string const &json)
+  {
+    try {
+      auto r {write(json)};
+      auto applied_all = std::all_of(r.begin(), r.end(), [](auto const &result){ return result == consensus::apply_ret_t::APPLIED; });
+      if (!applied_all)
+        {
+          throw std::runtime_error("This didn't work: " + json);
+        }
+      ASSERT_TRUE(applied_all);
+    }
+    catch (std::exception &ex) {
+      throw std::runtime_error(std::string(ex.what()) + " processing " + json);
+    }
+  }
+
+  void assertEqual(std::shared_ptr<velocypack::Builder> result, std::string const &expected_result) const {
+    try {
```

Comment:
Again: Why `try` if you throw again anyway?

Reply:
Because I am adding contextual information to whatever exception was generated on the called functions. Makes the error report a lot more useful.

Reply:
ok